core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,491 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.log;
+
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * {@link org.apache.logging.log4j.message.ParameterFormatter}
+ */
+public class MessageFormat {
+  /**
+   * Prefix for recursion.
+   */
+  static final String RECURSION_PREFIX = "[...";
+  /**
+   * Suffix for recursion.
+   */
+  static final String RECURSION_SUFFIX = "...]";
+
+  /**
+   * Prefix for errors.
+   */
+  static final String ERROR_PREFIX = "[!!!";
+  /**
+   * Separator for errors.
+   */
+  static final String ERROR_SEPARATOR = "=>";
+  /**
+   * Separator for error messages.
+   */
+  static final String ERROR_MSG_SEPARATOR = ":";
+  /**
+   * Suffix for errors.
+   */
+  static final String ERROR_SUFFIX = "!!!]";
+
+  private static final char DELIM_START = '{';
+  private static final char DELIM_STOP = '}';
+  private static final char ESCAPE_CHAR = '\\';
+
+  private static ThreadLocal<SimpleDateFormat> threadLocalSimpleDateFormat = new ThreadLocal<>();
+
+  private MessageFormat() {
+    // only static
+  }
+
+  /**
+   * Replace placeholders in the given messagePattern with arguments.
+   *
+   * @param messagePattern the message pattern containing placeholders.
+   * @param arguments      the arguments to be used to replace placeholders.
+   * @return the formatted message.
+   */
+  public static String format(final String messagePattern, @Nullable final Object[] arguments) {
+    final StringBuilder result = new StringBuilder();
+    final int argCount = arguments == null ? 0 : arguments.length;
+    formatMessage(result, messagePattern, arguments, argCount);
+    return result.toString();
+  }
+
+  /**
+   * Replace placeholders in the given messagePattern with arguments.
+   *
+   * @param buffer the buffer to write the formatted message into
+   * @param messagePattern the message pattern containing placeholders.
+   * @param arguments      the arguments to be used to replace placeholders.
+   */
+  public static void formatMessage(final StringBuilder buffer, @Nullable final String messagePattern,
+    @Nullable final Object[] arguments, final int argCount) {
+    if (messagePattern == null || arguments == null || argCount == 0) {
+      buffer.append(messagePattern);
+      return;
+    }
+    int escapeCounter = 0;
+    int currentArgument = 0;
+    int i = 0;
+    final int len = messagePattern.length();
+    for (; i < len - 1; i++) { // last char is excluded from the loop
+      final char curChar = messagePattern.charAt(i);
+      if (curChar == ESCAPE_CHAR) {
+        escapeCounter++;
+      } else {
+        if (isDelimPair(curChar, messagePattern, i)) { // looks ahead one char
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Move this trailing comment on the previous empty line. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)